keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md

```diff
@@ -125,29 +124,99 @@ serves the resourceVersion="0" list requests from reflectors today.
-  Use the [getCurrentResourceVersionFromStorage] added as part of [Watch-List KEP].
-- If the cache already has the current revision, serve the request from cache. If not,
-  - Send a `WatchProgressRequest` to etcd on the watch channel that the watch cache is consuming.
-- Use the existing `waitUntilFreshAndBlock` function in the watch cache to wait briefly for the watch to catch up to the current revision.
-- If the block times out, the request will result in rejection. (see "What if the watch cache is stale?" section for details)
+- Check if the watch cache already has the current revision, if not:
+  - Add a "waiting read" and notify the goroutine running in background.
+  - Wait for watch cache to catch up. If wait times out, reject the request.
+    (see "What if the watch cache is stale?" section for details)
+  - Remove a "waiting read"
+- Serve the request from watch cache
+
+In background, run a dedicated goroutine that will:
+- Wait for changes to "waiting read" count.
+- Repeat as long there is at least one "waiting read":
+  - Send `WatchProgressRequest` request to etcd.
+  - Wait 100ms
 
 Consistent GET requests will continue to be served directly from etcd. We will
 only serve consistent LIST requests from cache.
 
-[getCurrentResourceVersionFromStorage]: https://github.com/kubernetes/kubernetes/blob/3f247e59edfd4083242ad7271d076a38291760ff/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L1246-L1278
 [Watch-List KEP]: /keps/sig-api-machinery/3157-watch-list
 
+### Bug in etcd progress notification
+
+Only recently community discovered a bug [etcd-io/etcd#15220] for requesting
+progress notification. This bug causes a race between sending an event and
+progress notification with the same revision. Normally we would expect to event
+to always come first, however due to the bug, progress notification might be
+sent first. Serving a consistent lists in that situation could result in
+returning too early and missing the event. The bug was only fixed in v3.4.25 and
+v3.5.8 (both 9 months old), which is too fresh for Kubernetes to not handle.
+
+Our tests have shown that with the bug manual progress notification cannot be
+trusted at all. It causes a silent corruption that cannot be automatically
+detected prior to acting upon the corrupted data. For Beta, we propose to
+conditionally enable the feature in a way that mitigates the risk of unaware
+users stumbling on the bug.
+
+We propose to implement a safeguard that will prevent enabling
+`ConsistentListFromCache` on etcd version with broken progress notification.
+During kube-apiserver startup we will verify etcd version by calling `Status`
+method and checking etcd patch version. The condition will only pass if all etcd
+endpoints have etcd version that includes the fix for [etcd-io/etcd#15220].
+
+The proposed behavior of the safeguard for Beta:
+* etcd version could not be acquired - warn and disable the feature
```

Comment:
I think in this case, both here and for GA, we should 

```suggestion
* etcd version could not be acquired - warn, but continue with whatever the current value of the featuregate is
```

If someone uses not-etcd (kine?), I don't have a desire to break them or add recognition for their versions. They'll get the default server behavior and have to uphold the etcd contract.

Reply:
Done.